csharp/src/SeedLang/Runtime/Events.cs

```diff
@@ -12,24 +12,49 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+using System.Diagnostics;
 using SeedLang.Common;
 
 namespace SeedLang.Runtime {
+  public abstract class AbstractEvent {
+    // The range of source code.
+    public Range Range { get; }
+
+    public AbstractEvent(Range range) {
+      Range = range;
+    }
+  }
+
   // An event which is triggered when a binary expression is evaluated.
-  public class BinaryEvent {
+  public class BinaryEvent : AbstractEvent {
     public IValue Left { get; }
     public BinaryOperator Op { get; }
     public IValue Right { get; }
     public IValue Result { get; }
-    // The source code range of the binary expression.
-    public Range Range { get; }
 
-    public BinaryEvent(IValue left, BinaryOperator op, IValue right, IValue result, Range range) {
+    public BinaryEvent(IValue left, BinaryOperator op, IValue right, IValue result, Range range) :
+        base(range) {
       Left = left;
       Op = op;
       Right = right;
       Result = result;
-      Range = range;
+    }
+  }
+
+  // An event which is triggered when a compare expression is evaluated.
```

Comment:
Naming: `Comparison expression`

Here and others

Reply:
Done